### DISCUSSION TITLE
Error with Qt moc when using VCPkg manifests/registries

### DISCUSSION THREAD

**Discussion by drathin4** [microsoft/vcpkg#27636] [Q&A]
I'm trying to add a custom library to `VCPkg` using the `manifests/registries` feature on `Windows` platform. I have been in in touch with the support team regarding this and have initiated discussions: #27573, #27308, #27528, #27500.

Currently I'm facing an issue with the `Qt` dependency of the library `fleck` that I want to integrate. This library builds successfully when using `classic` mode of `VCPkg` using the `CMake` build commands (after removing the `PCL` dependency - #27573). But when I use `VCPkg manifests/registries` to integrate it, I get the **linker errors** connected to **Q_OBJECT**. There are some instances where the macro **Q_OBJECT** has been used in *.h files. 

5>mo_fleckwin.obj : error LNK2001: unresolved external symbol "public: static struct QMetaObject const Mo_Mural::staticMetaObject" (?staticMetaObject@Mo_Mural@@2UQMetaObject@@B) [D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\fleck.vcxproj]
5>D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\bin\Debug\fleck.exe : fatal error LNK1120: 1 unresolved externals [D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\fleck.vcxproj]
5>Done Building Project "D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\fleck.vcxproj" (default targets) -- FAILED.
3>Done Building Project "D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\ALL_BUILD.vcxproj" (default targets) -- FAILED.
1>Done Building Project "D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\install.vcxproj" (default targets) -- FAILED.

From the linker error, it seems to me that `moc` is not getting used despite of writing `set(CMAKE_AUTOMOC ON)` in `CMakeLists.txt`. I’m attaching that file and also `portfile.cmake` for your reference. 
[CMakeLists.txt](https://github.com/microsoft/vcpkg/files/9930291/CMakeLists.txt)
[portfile.cmake.txt](https://github.com/microsoft/vcpkg/files/9930310/portfile.cmake.txt)

The custom library (point cloud engine) has 4 sub-libraries of which `montage` is the one using `Qt`. The 4 places where **Q_OBJECT** macro is used: `montage/collage/mo_fleckwin.h, montage/collage/mo_toolbar.h, montage/mural/mo_mural.h, montage/mural/mo_treemodel.h`. Do I need to make any changes in `CMakeLists.txt or portfile.cmake (vcpkg_cmake_configure)` ?

Also FYI, I had to make 1 change in `CMakeLists.txt` when using `manifests/registries` as opposed to `classic` mode. I had to write `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)` to create an `import library (mural.lib)`. So I guess some changes take place inside `VCPkg` when using `manifests/registries`. Do we need to explicitly state them in `portfile.cmake (vcpkg_cmake_configure)` ?

**Comment by Neumann-A** [comment]
> From the linker error, it seems to me that moc is not getting used despite of writing set(CMAKE_AUTOMOC ON) in CMakeLists.txt

no. Check the binary directories if the `moc_` files are generated. Typically that linker error means you forgot to `#include "moc_filename.cpp"` in your `filename.cpp`. 

> when using manifests/registries as opposed to classic mode.

That is nonsense. classic and manifest mode behave exactly the same if the same baseline is used (same baseline in classic mode means the `ports/` folder is the checkout of said baseline). 

> set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON) to create an import library (mural.lib)

that means you are not correctly exporting symbols via `__declspec(dllexport)`

**Reply by drathin4** [reply]
I've written `#include "moc_filename.cpp"` in `filename.cpp` for all the files, but despite that I'm getting the same error:

fleck.dir\Debug\mo_include.obj
fleck.dir\Debug\mo_appmain.obj
fleck.dir\Debug\mo_fleckwin.obj
fleck.dir\Debug\mo_fleckdoc.obj
fleck.dir\Debug\mo_canvas.obj
fleck.dir\Debug\mo_jsondoc.obj
fleck.dir\Debug\qrc_mo_resources.obj
5>`mo_fleckwin.obj : error LNK2001: unresolved external symbol "public: static struct QMetaObject const Mo_Mural::staticMetaObject" (?staticMetaObject@Mo_Mural@@2UQMetaObject@@B) [D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\fleck.vcxproj]`
5>`D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\bin\Debug\fleck.exe : fatal error LNK1120: 1 unresolved externals [D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\fleck.vcxproj]`
5>Done Building Project "D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\fleck.vcxproj" (default targets) -- FAILED.
3>Done Building Project "D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\ALL_BUILD.vcxproj" (default targets) -- FAILED.
1>Done Building Project "D:\dev\vcpkg\buildtrees\fleck\x64-windows-dbg\install.vcxproj" (default targets) -- FAILED.

Is it because `__declspec(dllexport)` is not mentioned in the source code ?

Please refer to `CMakeLists.txt` file for the next discussion. As per the log shown above, the `mo_*.obj` files correspond to `FL_COLLAGE_SOURCES` which is used in `add_executable(fleck ${FL_COLLAGE_SOURCES})`. Also `mo_fleckwin.obj` is complaining about `Mo_Mural` which is actually a part of `mo_mural.cpp`. This `mo_mural.cpp` is included in `FL_MURAL_SOURCES` which forms the dynamic library `mural`. In that context, where do I have to use `__declspec(dllexport)`? 

As per the source code files, I couldn't find any connection between `mo_mural.cpp` and `mo_fleckwin.cpp`. Also in `CMakeLists.txt`, `mural` library is already linked to `fleck` executable. So, where exactly to `export/import` the symbols ?

**Comment by JavierMatosD** [comment]
Hi @drathin4, 

Would you mind providing/answering the following: 
1. The steps you took to build your library successfully in classic mode with the commands used.
2. What do you mean by  "use VCPkg manifests/registries to integrate it"? Are you trying to build your library in manifest mode? 
3. The steps you took to integrate your library.

Thank you,
Javi

**Reply by drathin4** [reply]
I have removed Qt dependency from the library for the time being. If I face the same issue after I include the dependency again, I will write in this thread. Thanks.
